pkg/autoscaling/metrics/prometheus.go

```diff
@@ -0,0 +1,210 @@
+	tidbSumCPUUsageMetricsPattern = `sum(increase(process_cpu_seconds_total{cluster="%s",job="tidb"}[%s])) by (instance)`
+	tikvSumCPUQuotaMetricsPattern = `tikv_server_cpu_cores_quota{cluster="%s"}`
+	tidbSumCPUQuotaMetricsPattern = `tidb_server_maxprocs{cluster="%s"}`
+	queryPath                     = "/api/v1/query"
+	statusSuccess                 = "success"
+
+	httpRequestTimeout = 5
+)
+
+// Response is used to marshal the data queried from Prometheus
+type Response struct {
+	Status string `json:"status"`
+	Data   Data   `json:"data"`
+}
+
+// Data consists of response data from prometheus
+type Data struct {
+	ResultType string   `json:"resultType"`
+	Result     []Result `json:"result"`
+}
+
+// Result consists of value and its labels
+type Result struct {
+	Metric Metric        `json:"metric"`
+	Value  []interface{} `json:"value"`
+}
+
+// Metric consists of labels
+type Metric struct {
+	Cluster  string `json:"cluster,omitempty"`
+	Instance string `json:"instance"`
+	Job      string `json:"job,omitempty"`
+}
+
+// PrometheusQuerier query metrics from Prometheus
+type PrometheusQuerier struct {
+	// Prometheus API Endpoint Address
+	endpoint string
+	client   promClient.Client
+}
+
+// NewPrometheusQuerier returns a PrometheusQuerier
+func NewPrometheusQuerier(endpoint string) (*PrometheusQuerier, error) {
+	client, err := promClient.NewClient(promClient.Config{Address: endpoint})
+	if err != nil {
+		return nil, err
+	}
+
+	store := &PrometheusQuerier{
+		endpoint,
+		client,
+	}
+
+	return store, nil
+}
+
+type promQLBuilderFn func(*QueryOptions) (string, error)
+
+var queryBuilderFnMap = map[MetricType]promQLBuilderFn{
+	CPUQuota: buildCPUQuotaPromQL,
+	CPUUsage: buildCPUUsagePromQL,
+}
+
+// Query do the real query on Prometheus and returns metric value for each instance
+func (prom *PrometheusQuerier) Query(options *QueryOptions) (QueryResult, error) {
+	builderFn, ok := queryBuilderFnMap[options.metric]
+	if !ok {
+		return nil, errors.Errorf("unsupported metric type %v", options.metric)
+	}
+
+	query, err := builderFn(options)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := prom.queryMetricsFromPrometheus(query, options.timestamp)
+	if err != nil {
+		return nil, err
+	}
+
+	result, err := extractInstancesFromResponse(resp, options.instances)
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func (prom *PrometheusQuerier) queryMetricsFromPrometheus(query string, timestamp int64) (*Response, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*httpRequestTimeout)
+	defer cancel()
+	req, err := http.NewRequestWithContext(ctx, "GET", fmt.Sprintf("%s%s", prom.endpoint, queryPath), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	q := req.URL.Query()
+	q.Add("query", query)
+	q.Add("time", fmt.Sprintf("%d", timestamp))
+	req.URL.RawQuery = q.Encode()
+	r, body, _, err := prom.client.Do(req.Context(), req)
```

Comment:
Do we need to close this body?

Reply:
No, the client is responsible for closing body. And the body here is a []byte